### PR TITLE
Annotate callback passed to C function to match upstream

### PR DESCRIPTION
**Pull request by mathias-lang-sociomantic**
Since delegate parameter contravariance is not implemented, this error out when compiled with upstream compiler,
which annotate the parameter of the callback with scope.
With our version of dmd(-transitional), it doesn't make a difference though.
Moreover, thanks to the prototype change, some implicit convertion is not done anymore.

Part of #9

### CODE REVIEW

**Comment by mihails-strasuns-sociomantic** [comment]
Is there any chance we can fix `scope` delegates to not be checked in `@system` code instead, same as rest of `scope`? I am so not happy about dealing with any of dip1000 consequences >_<

**Comment by mathias-lang-sociomantic** [comment]
> Is there any chance we can fix scope delegates to not be checked in @system code instead, same as rest of scope? I am so not happy about dealing with any of dip1000 consequences >_<

Yeah me neither, but the issue is not `scope` related. Since delegate parameter contravariance is not implemented, there's not much we can do. And implementing it is quite a project.

IOW, since this code doesn't compile:
```D
void foo (void function(mstring) dg);
void bar (cstring);
foo(&bar); // Error !?!
````

This code cannot possibly compile without hacks:
```D
void foo (void function(scope void*) dg);
void bar (void*);
foo(&bar); // Ditto error
```

**Comment by mihails-strasuns-sociomantic** [comment]
Ah, right, so the problem boils down to extra `scope` added in `druntime` to C functions args?

**Comment by mihails-strasuns-sociomantic** [comment]
I wonder if redefining the binding as private symbol in this module would be less of a mess in that case :(

**Comment by mathias-lang-sociomantic** [comment]
> Ah, right, so the problem boils down to extra `scope` added in druntime to C functions args?

Yes

> I wonder if redefining the binding as private symbol in this module would be less of a mess in that case :(

Hum no ? We'll eventually have to comply with some of those things, so it just sounds like extra work. This change is not *wrong* per se. Things would be smoother if the type system was better, but I don't see a problem in adding this annotation *for this case*.

File: src/ocean/io/select/selector/TimeoutSelectedKeysHandler.d [hunk 1]
```diff
@@ -195,8 +197,9 @@ class TimeoutSelectedKeysHandler: SelectedKeysHandler
 
     ***************************************************************************/
 
-    extern (C) private static int cmpPtr ( bool searching ) ( in void* a_,
-        in void* b_ )
+    extern (C) private static int cmpPtr ( bool searching ) (
+        /* d1to2fix_inject: scope */ Const!(void*) a_,
```
Comment (david-eckardt-sociomantic) [inline src/ocean/io/select/selector/TimeoutSelectedKeysHandler.d:201]: `in void* a_`?

Comment (mihails-strasuns-sociomantic) [reply src/ocean/io/select/selector/TimeoutSelectedKeysHandler.d:201]: Can't mix `in` and `scope` because former was supposed to include the latter (but not anymore). Hm, speaking about that, @mathias-lang-sociomantic do you think this restriction can/should be removed now?

Comment (mathias-lang-sociomantic) [reply src/ocean/io/select/selector/TimeoutSelectedKeysHandler.d:201]: Which restriction ? That `in` doesn't include `scope` ? I would love that, but don't think Walter will agree. Also, [they don't play along very well](https://issues.dlang.org/show_bug.cgi?id=17408) :(

Comment (mihails-strasuns-sociomantic) [reply src/ocean/io/select/selector/TimeoutSelectedKeysHandler.d:201]: Other way around, prohibiting `in scope` combo.

Comment (mathias-lang-sociomantic) [reply src/ocean/io/select/selector/TimeoutSelectedKeysHandler.d:201]: Should the bug be fixed ? Definitely. Just didn't have  the time (nor considered it important enough, given how DIP1000 is broken) to get around it.

Comment (mihails-strasuns-sociomantic) [reply src/ocean/io/select/selector/TimeoutSelectedKeysHandler.d:201]: I wasn't even sure it is viewed as bug now. Was definitely intentional when `in` was still supposed to mean `const scope`.

File: src/ocean/io/select/selector/TimeoutSelectedKeysHandler.d [hunk 2]
```diff
@@ -143,7 +144,8 @@ class TimeoutSelectedKeysHandler: SelectedKeysHandler
 
                     verify(client !is null);
 
-                    if (!bsearch(cast (void*) client, timed_out_clients.ptr,
```
Comment (mihails-strasuns-sociomantic) [inline src/ocean/io/select/selector/TimeoutSelectedKeysHandler.d:146]: Why is this change necessary btw?

Comment (mathias-lang-sociomantic) [reply src/ocean/io/select/selector/TimeoutSelectedKeysHandler.d:146]: Because the second parameter is `inout(void*)` but we pass a typed parameter (`Timeout*` IIRC). Apparently DMD cannot figure out the conversion.

**Review by mihails-strasuns-sociomantic — APPROVED** [review]
